### PR TITLE
Update to protovalidate 0.6.1

### PR DESCRIPTION
**Pull request by pkwarren**
Update protovalidate-java to support the latest rules in protovalidate 0.6.1. Add the isHostAndPort overload and apply additional fixes from protovalidate-go to the Java implementation.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -153,9 +153,10 @@ tasks.withType<JavaCompile> {
 }
 
 tasks.withType<Javadoc> {
-    // TODO: Enable when Javadoc changes are final
-//    val stdOptions = options as StandardJavadocDocletOptions
-//    stdOptions.addBooleanOption("Xwerror", true)
+    val stdOptions = options as StandardJavadocDocletOptions
+    stdOptions.addBooleanOption("Xwerror", true)
+    // Ignore warnings for generated code.
+    stdOptions.addBooleanOption("Xdoclint/package:-build.buf.validate,-build.buf.validate.priv", true)
```
Comment (pkwarren) [inline build.gradle.kts:159]: Figured out how to ignore warnings on Javadoc for generated protoc code but still mark warnings as errors for our library code.

File: src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java [hunk 2]
```diff
@@ -438,6 +440,55 @@ private static Overload isInf() {
...
+            return Err.noSuchOverload(lhs, OVERLOAD_IS_HOST_AND_PORT, rhs);
+          }
+          String value = (String) lhs.value();
+          boolean portRequired = rhs.booleanValue();
+          return Types.boolOf(hostAndPort(value, portRequired));
+        },
+        null);
+  }
+
+  private static boolean hostAndPort(String value, boolean portRequired) {
```
Comment (pkwarren) [inline src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java:459]: Based off Go implementation.

File: src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java [hunk 3]
```diff
@@ -484,7 +535,7 @@ private static boolean validateEmail(String addr) {
     try {
       InternetAddress emailAddr = new InternetAddress(addr);
       emailAddr.validate();
-      if (addr.contains("<")) {
+      if (addr.contains("<") || !emailAddr.getAddress().equals(addr)) {
```
Comment (pkwarren) [inline src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java:538]: Upstream Go implementation got this fix for failing on whitespace in email addresses.

File: src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java [hunk 4]
```diff
@@ -510,7 +561,9 @@ private static boolean validateHostname(String host) {
     }
     String s = Ascii.toLowerCase(host.endsWith(".") ? host.substring(0, host.length() - 1) : host);
     Iterable<String> parts = Splitter.on('.').split(s);
+    boolean allDigits = false;
```
Comment (pkwarren) [inline src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java:564]: Upstream Go impl also picked up additional validation for hosts to ensure the last portion isn't all numbers.

File: src/main/java/build/buf/protovalidate/internal/evaluator/EvaluatorBuilder.java [hunk 5]
```diff
@@ -226,16 +233,40 @@ private FieldEvaluator buildField(
         FieldDescriptor fieldDescriptor, FieldConstraints fieldConstraints)
         throws CompilationException {
       ValueEvaluator valueEvaluatorEval = new ValueEvaluator();
+      boolean ignoreDefault =
```
Comment (pkwarren) [inline src/main/java/build/buf/protovalidate/internal/evaluator/EvaluatorBuilder.java:236]: Updates here are mostly equivalent to https://github.com/bufbuild/protovalidate-go/compare/v0.5.2...2268789d9be1f269f0001d64f5dd1bdd60278964#diff-70b43cf25409483d9785a2bea15d9d185795803e255330e6562e7dca5112663e (there are some differences in the Java vs. Go generated code).

File: src/main/java/build/buf/protovalidate/exceptions/ExecutionException.java [hunk 6]
```diff
@@ -16,6 +16,11 @@
 
 /** ExecutionException is returned when a constraint fails to execute. This is a fatal error. */
 public class ExecutionException extends ValidationException {
+  /**
+   * Creates an executioon exception with the specified message.
```
Comment (drice-buf) [inline src/main/java/build/buf/protovalidate/exceptions/ExecutionException.java:20]: Typo: `executioon` -> `execution`

File: src/main/java/build/buf/protovalidate/exceptions/ValidationException.java [hunk 7]
```diff
@@ -16,10 +16,21 @@
 
 /** ValidationException is the base exception for all validation errors. */
 public class ValidationException extends Exception {
+  /**
+   * Creates a new ValidationException with the specified message.
```
Comment (drice-buf) [inline src/main/java/build/buf/protovalidate/exceptions/ValidationException.java:20]: Super-nit: use a consistent phrasing (right now some places have "Creates a ..." and other places have "Creates a new ...")

File: src/main/java/build/buf/protovalidate/exceptions/ExecutionException.java (hunk 6 above)
Comment (pkwarren) [reply src/main/java/build/buf/protovalidate/exceptions/ExecutionException.java]: Fixed.

File: src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java [hunk 8]
```diff
@@ -520,9 +573,11 @@ private static boolean validateHostname(String host) {
         if ((ch < 'a' || ch > 'z') && (ch < '0' || ch > '9') && ch != '-') {
```
Comment (drice-buf) [inline src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java:573]: Some small helper functions can make this more readable like `if (!isAlpha(ch) && !isDigit(cg) && ch != '-')`

**Review by drice-buf — APPROVED** [review]
LGTM with some small comments/nits

File: src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java (hunk 8 above)
Comment (pkwarren) [reply src/main/java/build/buf/protovalidate/internal/celext/CustomOverload.java]: Fixed.

File: src/main/java/build/buf/protovalidate/internal/evaluator/EvaluatorBuilder.java [hunk 9]
```diff
@@ -226,16 +232,40 @@ private FieldEvaluator buildField(
...
+          fieldDescriptor.hasPresence() && shouldIgnoreDefault(fieldConstraints);
+      Object zero = null;
```
[the remark this answers is not in the document]
Comment (pkwarren) [reply src/main/java/build/buf/protovalidate/internal/evaluator/EvaluatorBuilder.java:237]: Didn't make this change - mainly a style thing and I think the ternary operator can make things harder to follow. The only use of the ternary operator so far in this project is in the protoc generated code.

**Review by drice-buf — APPROVED** [review]

**Pull request merged**